### ISSUE TITLE
Implement Traversable.slideBy(classifier)

### ISSUE THREAD

**Issue by danieldietrich** [javaslang/javaslang#1775]
**Example:**

```java
// Vector(List(a1, a2), List(b1), List(a3))
List("a1", "a2", "b1", "a3").slideBy(s -> s.substring(0, 1))
```

See also [Scala List function for grouping consecutive identical elements](http://stackoverflow.com/questions/4761386/scala-list-function-for-grouping-consecutive-identical-elements)

---

**Traversable** has

> _// same as sliding(size, size)_
> [grouped(int size)](http://static.javadoc.io/io.javaslang/javaslang/2.1.0-alpha/javaslang/collection/Traversable.html#grouped-int-) &rightarrow; Iterator&lt;? extends Traversable&lt;T&gt;&gt;
> [groupBy(Function&lt;? super T,? extends C&gt; classifier)](http://static.javadoc.io/io.javaslang/javaslang/2.1.0-alpha/javaslang/collection/Traversable.html#groupBy-java.util.function.Function-) &rightarrow; Map&lt;C,? extends Traversable&lt;T&gt;&gt;

> _// same as sliding(size, 1)_
> [sliding(int size)](http://static.javadoc.io/io.javaslang/javaslang/2.1.0-alpha/javaslang/collection/Traversable.html#sliding-int-) &rightarrow; Iterator&lt;? extends Traversable&lt;T&gt;&gt;
> [sliding(int size, int step)](http://static.javadoc.io/io.javaslang/javaslang/2.1.0-alpha/javaslang/collection/Traversable.html#sliding-int-int-) &rightarrow; Iterator&lt;? extends Traversable&lt;T&gt;&gt;

The new function will be also part of Traversable, namely:

> **slideBy(Function&lt;? super T,? extends C&gt; classifier) &rightarrow; Seq&lt;? extends Traversable&lt;T&gt;&gt;**

Thoughts on the implementation:

* The returned Seq is a Vector (internally), as part of our strategy to make Vector the default Seq
* We should also implement it as part of Collections (see Collections.groupBy).
* There will be no default implementation. The methods need to be overridden throughout the collection hierarchy (see Traversable.groupBy()), i.e.

> Traversable.slideBy(classifier) &rightarrow; Seq&lt;? extends Traversable&gt;
> Seq.slideBy(classifier) &rightarrow; Seq&lt;? extends Seq&gt;
> LinearSeq.slideBy(classifier) &rightarrow; Seq&lt;? extends LinearSeq&gt;
> List.slideBy(classifier) &rightarrow; Seq&lt;List&gt;

Please note that the 'terminal' types (like List, not necessarily a final class) do not specify bounded generics as part of the return type.

**Comment by jest** [comment]
I'll take a look at it over the weekend.

One question: I still have problems choosing between Iterator and `Seq` implementations, probably due to my use cases (huge amounts of short-lived objects); what is the rationale of making the new `slideBy` return `Seq` instead of `Iterator`, as `group*` and `sliding` do (or maybe the other way round, why `Iterator` instead of `Seq`?) Behind the scene there will be a `Vector` anyway, so either return types are "easy" enough from the point of view of implementation.

**Comment by danieldietrich** [comment]
@jest yes, this is a valid question.

**TL;DR:**

Let's consider the return type `Iterator`:

```java
Iterator<? extends Traversable<T>> slideBy(Function<? super T,? extends C> classifier);
```

If possible, we should implement the Iterator directly (like Iterator.sliding(int, int)) but not performing an internal conversion to Stream.

Other collection.slideBy() implies can than simply call

```java
iterator().slideBy(classifier).map(CollectionType::ofAll);
```

---

I'm an observer, I looked at Scala's API and asked myself that question several times before.

Generally it is a good idea to consume time/mem when there is a demand for a computation. As we see, most javaslang.collection.Iterator methods can be implemented in a lazy way. So why the (non-reducing) operations return a new collection instead of Iterator. And, on the other hand, why there exist operations (like grouped and sliding) that return an Iterator?

We do not need the Iterator type to reflect laziness. All collections could be implemented in a lazy way (under the hood).

I think there are two main points returning a collection type (like filter and map do):

* The type reflects specific characteristics/constraints that hold. E.g. when we filter the elements of a Set, the resulting subset is still a Set (= distinct, finite and not sequential).
* A collection can be traversed again (starting with 3.0.0 Iterator will not implement Traversable any more).

On the other hand there these are the reasons to return an Iterator:

* Iterative results may save intermediate object creation, they may save time and mem.
* A result of type Iterator may be preferred over Seq if the only characteristics of the result is _sequential_ **and** most probably no random access is needed.

However, if we decide to return Iterator, we can be still obtain another collection instance in O(n) by calling one of our conversion methods.

**In the case of slideBy() I see no difference to sliding(), so we should return an Iterator.**

**Comment by jest** [comment]
OK, I'm working on it; as it is similar to `sliding()`, which is implemented in most of classes as `iterator().sliding(...).map(...)`, I started with `Iterator.sliding()`, although I don't understand what you mean by "but not performing an internal conversion to Stream" in your comment.

As for the method signature, is `C` type variable necessary? It is not used anywhere in the signature, and the classifier results are compared using `Object.equals()`. So maybe such a signature is enough?

```
Iterator<? extends Traversable<T>> slideBy(Function<? super T,?> classifier);
```

**Comment by danieldietrich** [comment]
@jest alright!

> although I don't understand what you mean by "but not performing an internal conversion to Stream" in your comment.

I looked at the implementation of Iterator.sliding() and recognised that we use split() and drop(), which need to traverse the source twice. Therefore the source Iterator is converted to a Stream, which can be traversed more than once. I think a different implementation is possible (using some state variables) without conversion to a Stream.

> As for the method signature, is C type variable necessary?

You are right, the signature you suggested should perfectly do the job. I think `?` is sufficient (compared to an alternative like `? extends Object`).

Thank you!

**Comment by jest** [comment]
Man, that is a whole lot of file editing for such a small feature. :-) Anyway, first result is in https://github.com/jest/javaslang/tree/slideby-feature. Initially naively copying `Iterator.sliding()` with small tweaks, and no tests.

**Comment by danieldietrich** [comment]
Thank you, looks great!

> Man, that is a whole lot of file editing for such a small feature. :-)

Ohhh yeah, that's true. That's why I want to generate all this redundant boilerplate automatically. See #1716

**Comment by jest** [comment]
OK, I'll add more tests in ~12 hours and issue a PR

**Comment by danieldietrich** [comment]
Thank you!

**Comment by jest** [comment]
OK, added tests. They fail however on `HashSetTest` due to "random" iteration order. Not sure how to warn a user about it (JavaDoc?) The same applies to `sliding`.

**Comment by danieldietrich** [comment]
Yes. Interesting that we seem to rely on the order of sliding() in HashSetTest. Anyway - we should override the non-functional tests in HashSetTest (with an empty implementation). Maybe you could add a `of(1, 2, 3).slideBy(Function.identity())`-test (in AbstractTraversableTest) that asserts that the result equals `Vector(of(1), of(2), of(3))` without taking the order into account (should be already the case because we override `assertThat(Iterable<T> actual)`, see HashSetTest).

**Comment by jest** [comment]
This is already there, https://github.com/jest/javaslang/commit/9e772804aeded7e74c614c4c81e226ba9d33cd82#diff-7e6e25e07a7b892583cb2ad871a61b53R1554, and yes, as `assertThat` is non-static, it is used from `HashSetTest` in its tests.

**Comment by danieldietrich** [comment]
Great!

---

Is it (generally) really necessary to convert test-results to Vector (using `map(Vector::ofAll)`)? The tests could look like this:

```java
@Test
public void shouldSlideNotNilByIdentityClassifier() {
    final List<Traversable<Integer>> actual = of(1, 2, 3).slideBy(Function.identity()).toList();
    final List<Traversable<Integer>> expected = List.of(of(1), of(2), of(3));
    assertThat(actual).isEqualTo(expected);
}
```

**Comment by jest** [comment]
Not sure, just copy-pasted from `sliding()` tests and wanted them to look consistently

**Comment by danieldietrich** [comment]
Ok, it might make sense for Iterator.slideBy() / sliding(). This will change with 3.0.0. For now it is ok then!
